Scons windows

ericmehl opened the pull request:
Per the end of the discussion around #916 these changes add Scons building compatibility on Windows. There are quite a few changes that needed to be made so I've split them up thematically to hopefully make it easier to review and explain why the change was made.

This also adds Appveyor support. It currently downloads the dependencies from [my work-in-progress Gaffer Dependencies releases](https://github.com/hypothetical-inc/gafferDependencies/releases/tag/0.53.0.1).

Note that the following modules have not been tested to build on Windows:
- Maya
- Houdini
- Nuke

I've made the obvious changes to their build sections to keep them consistent with the other modules, but actual builds have not been tested.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/ImageEngine/cortex/blob/master/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [ ] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Cortex project's prevailing coding style and conventions.
- [X] If my code made breaking changes, I applied the **pr-majorVersion** label to this PR.

ericmehl commented:
It seems to be failing on only one configuration, I'll close and reopen to trigger a new build.

boberfly commented:
Oh wow you got it to work with SCons, nice! Is it faster than CMake?

ericmehl commented:
It's about the same speed on the first build as Cmake (using JOM) but the nice thing is that subsequent builds are a lot faster since it can use the Scons cache and only rebuild what's needed, even on a clean VM that Appveyor fires up for a build.

The initial build was coming in at something like 50min and it's gone down to at least 14min after that. I suspect it will go lower too - with all the pushes I was doing I'm not exactly sure what was cached before that 14min build.

johnhaddon commented on appveyor.yml:
```diff
@@ -0,0 +1,133 @@
+#---------------------------------#
+
+# version format
+version: 0.53.0.0.{build}
```
This looks like a Gaffer version, rather than a Cortex one, and the same goes for the VERSION environment variable below. I'm also wondering if we can do something to derive them automatically from the contents of the SConstruct file, because I fear they'll quickly get out of sync otherwise.

ericmehl replied:
Ah yes, good catch. I'll see if I can modify the build number from a script in the build process.

boberfly commented:
Hey John,

The beauty of clang-cl is that it is a drop-in replacement of MSVC so it is actually pretending it is MSVC itself so these flags should just work, much like how clang can emulate GCC. The only thing I recommend is that we shouldn't imply that the compiler host is running on Windows but could be cross-compiled on macOS/Linux as well. I was playing with the build.py stuff in the dependencies repo where you can specify your cross-compile target for this, but it looks like it'll be quite a bit of time and effort in fixing and testing for the dependencies to not imply the host is Windows so I dropped that for now :(

johnhaddon commented:
> The beauty of clang-cl is that it is a drop-in replacement of MSVC so it is actually pretending it is MSVC itself so these flags should just work, much like how clang can emulate GCC. 

Wouldn't you still need to specify `CXX=clang-cl` in that scenario though? My suggestion was to preserve the CXX argument so this was still possible...

ericmehl commented:
> p.s. Brilliant to see that the Appveyor setup seems so straightforward after this. Presumably we need to do something to get the main Cortex repo signed up for PR checks?

Yes, it works great! I believe it's as simple as someone creating a free account and then signing in with their GitHub credentials. It differentiates between the individual's GitHub repos and the organizations they are part of, so it probably needs to be someone who is part of the Image Engine organization.

I think it took a couple steps to get it to recognize the Organization you are part of and add the project to the list Appveyor monitors, but after that it figures everything out.

All of the config lives in the appveyor.yml file in the repository so there isn't any configuration needed for the project itself.

ericmehl commented on SConstruct:
```diff
@@ -1587,6 +1587,8 @@ corePythonScripts = glob.glob( "python/IECore/*.py" )
 if doConfigure :
 
 	freetypeEnv = coreEnv.Clone()
+	freetypeEnv.Replace( CXXFLAGS = [i for i in freetypeEnv["CXXFLAGS"] if i != "-DIECore_EXPORTS"])
```
Good point, I like that better too. For some reason I had thought the CoreEnv was doing more setup work that would need to be repeated for each module but after looking it over again I see that isn't the case.

johnhaddon commented:
> Yes, it works great! I believe it's as simple as someone creating a free account and then signing in with their GitHub credentials. It differentiates between the individual's GitHub repos and the organizations they are part of, so it probably needs to be someone who is part of the Image Engine organization.

Cool - I'll take care of that part then. It seemed easy enough to set it up on my personal repo, so once we're ready should be easy to add it to the main ImageEngine repo too...

ericmehl commented on SConstruct:
```diff
@@ -1735,6 +1744,8 @@ if doConfigure :
 		if "-DIECORE_WITH_FREETYPE" in imageEnv["CPPFLAGS"] :
 			imageEnv.Append( LIBS = "freetype" )
+			if env["PLATFORM"] == "win32":
+				imageEnv.Append( LIBS = ["libpng"] )
```
It looks like FreeType uses PNG to "Support PNG compressed OpenType embedded bitmaps." I'm not too familiar with the fonts Gaffer and Cortex use but that sounds like something we could leave out. Do you agree?

Digging in the cmake for FreeType it turns out that if it can find the PNG library it will use it even if it's set to NO unless a CMAKE_DISABLE_FIND_PACKAGE is set for the PNG library.

I'll add that in the dependencies and that should remove the need for linking to PNG in Cortex.

johnhaddon replied:
Yes, I'd be very surprised if we needed PNG compressed OpenType embedded bitmaps. If you don't mind, let's give it a go without and see what happens.

ericmehl replied:
Cool, I sorted out the PNG linking in the dependencies so this commit is dropped now.

ericmehl commented on SConstruct:
```diff
@@ -1692,6 +1688,23 @@ imageEnvAppends = {
 	],
 }
 
+if env["PLATFORM"] != "win32" :
```
I agree it gets confusing, I'll take another look at more elegant options. I like your idea of setting $ISYSTEM - as long as setting that to -I for Windows and -isystem for others works I expect I can drop the CPPPATH business and keep it cleaner.

ericmehl replied:
I sorted this by adding a new function that does the conditional in one place rather than each module. Formatting technicalities made the env["ISYSTEM"] method not work. Hopefully a nice side effect of the function is that it makes it a little more explicit what is going on.

ericmehl commented on SConstruct:
```diff
@@ -1331,7 +1334,7 @@ if env["PLATFORM"] == "posix" :
 
 # get the python link flags
-if pythonEnv["PYTHON_LINK_FLAGS"]=="" :
+if pythonEnv["PLATFORM"] != "win32" and pythonEnv["PYTHON_LINK_FLAGS"]=="" :
```
Windows doesn't have python-config at all, but it also doesn't need any special flags to link correctly so it should be an empty string for Windows. Windows by default installs Python to C:\python27 (for example) but that can be overridden easily and in a dramatic twist it's up to the user to tell the installer to set the system-wide PATH variable to include their new Python installation.

So I think it's best to require the Python binary to be specified when building on Windows - I'll add a check for that and fail the build if it can't be found. I'll also clean this up some to make it more obvious what's happening on one platform vs the other.

ericmehl replied:
I cleaned this up a lot and added a new config test to make sure Python is working. Feel free to decline that addition if you think it's too much or it's causing problems.

ericmehl commented on SConstruct:
```diff
@@ -74,26 +76,27 @@ except NameError :
-	"The C++ compiler.",
-	"g++",
-)
+if Environment()["PLATFORM"] != "win32" :
```
Agreed, I'll put in a proper default to keep this clean.

andrewkaufman commented on SConstruct:
```diff
@@ -141,13 +144,13 @@ o.Add(
 	"BOOST_INCLUDE_PATH",
 	"The path to the boost include directory.",
-	"/usr/include",
+	"/usr/include" if Environment()["PLATFORM"] != "win32" else "",
```
I think that would make sense. I think the logic of the original defaults was to make things simpler for someone at home without an options file to build Cortex. But I suspect those folks are few and far between, especially now that gafferDependencies ships a precompiled version. I'm sure we're still inadvertently relying on one of the defaults at IE, but we may as well patch that up.

ericmehl replied:
I cleared out the defaults throughout in my latest commit. I found I needed to add VDB info to the Travis configuration as well.

ericmehl commented on SConstruct:
```diff
@@ -1184,7 +1184,7 @@ def getPythonConfig( env, flags ) :
 
 # decide where python is
-if pythonEnv["PYTHON"]=="" :
+if pythonEnv["PYTHON"]=="" and env["PLATFORM"] != "win32" :
```
This is sorted out with the larger reworking of the Python section in my new commit.

ericmehl commented on SConstruct:
```diff
@@ -1322,6 +1322,9 @@ else :
 pythonEnv.Append( CXXFLAGS = "$PYTHON_INCLUDE_FLAGS" )
 
+if pythonEnv["PLATFORM"] != "win32":
+	pythonEnv.Append( CXXFLAGS = "$PYTHON_INCLUDE_FLAGS" )
```
This is sorted out with the larger reworking of the Python section in my new commit.

ericmehl commented:
I reworked some big sections to clean things up. Interested to get your thoughts!

Travis ran out of time on one config so closing and reopening to kick it off again.

- Python checking section now makes the need for specifying the binary location on Windows explicit
- Included a config check for Python to ensure it's valid on Windows. It works on Linux too but if that creates more issues than it solves we can drop it
- Removed the conditional CPPPATH / -isystem switches throughout each module in favor of a new function includeSystemDirectory. This keeps it clean and puts the recurring conditional into that function. It wasn't as simple as adding a prefix to the CXXFLAGS due to formatting details (MSVC requires no space between the -I switch and the directory, gcc does need a space, and Scons automatically adds quotes around parameters with spaces that fouls up gcc)
- Removed the PNG linking and put up a new set of dependencies that doesn't require the PNG library in Freetype
- Added a new script in the config/appveyor directory to pull the version info from SConstruct and adapted appveyor to use that as it's version

ericmehl commented:
The latest force-push is just a rebase onto the latest to prevent slipping behind while under review. I added one commit at the end that recreates d5aa26f in the new framework of the Win32 build.

ericmehl commented:
Here's a fresh rebase on top of the latest changes. It still says 1 change requested but I think I addressed all of the old points. How's it looking now?

johnhaddon commented on SConstruct:
```diff
@@ -1281,6 +1281,11 @@ if doConfigure :
 		sys.stderr.write( "ERROR : unable to find the Blosc libraries - check BLOSC_INCLUDE_PATH and BLOSC_LIB_PATH.\n" )
 		Exit( 1 )
+	
+	if c.CheckLibWithHeader( "freetype", ["ft2build.h"], "CXX" ) :
```
Moving the freetype check here has the effect of adding `freetype` to the `LIBS` for `env`, causing _everything_ to link against freetype. Then intention was that only the libraries with a direct dependency would link it - for instance, IECoreImage does this :

```
if "-DIECORE_WITH_FREETYPE" in imageEnv["CPPFLAGS"] :
	imageEnv.Append( LIBS = "freetype" )
else :
	imageSources.remove( "src/IECoreImage/Font.cpp" )
	imagePythonSources.remove( "src/IECoreImageBindings/FontBinding.cpp" )
```

Can we put the call to `c.CheckLibWithHeader( "freetype", ["ft2build.h"], "CXX" )` back in its own `freetypeEnv` scope to avoid this "leak"? I think we can still put `-DIECORE_WITH_FREETYPE` in `env` as you have done here.

johnhaddon commented on SConstruct:
```diff
@@ -521,13 +521,13 @@ o.Add(
 	"VDB_INCLUDE_PATH",
 	"The path to the OpenVDB include directory.",
-	"/usr/local/include",
+	"",
```
This had a very "interesting" side effect for my builds, which took a while to track down. The symptom was a failure in the configure check for openvdb, with this error :

```
conftest_8.o: relocation R_X86_64_32S against symbol `_ZNKSt5ctypeIcE8do_widenEc' can not be used when making a shared object; recompile with -fPIC
```
It turned out that I had omitted to specify VDB_INCLUDE_PATH in my config, so what was `-isystem /usr/local/include` became just `-isystem `. And it just so happened that the argument after that was the all-important `-fPIC`, which was now being interpreted as an include path (`-system -fPIC`). This is totally my mistake, but it took me some time to figure out what was going in.

Can we make `append/prependSystemDirectory()` skip empty paths to avoid this pitfall? It'll be much easier to debug a "can't find header" error than a "relocation R_X86_64_32S against symbol" one.

mattigruener replied:
I'm late to the party, so please excuse if I'm getting the wrong end of the stick with any of this.

I've just done a test build with the setup used here at IE and I've run into a similar issue as you, John. Took me a little bit to hunt this down as well. 
Here at IE we do not specify `FREETYPE_INCLUDE_PATH`. I'm not sure why that's the case, but it causes our builds to lose the `-std=c++11` argument because of an `-isystem` that isn't followed by a path. The result was a whole bunch of errors regarding cpp11 keywords like `nullptr` and `override` etc.
Note that this is _with_ 1a6f3f4. I think that fix doesn't work because we don't need to check the entry in `include_list` but the path _after_ substitution instead. `"$FREETYPE_INCLUDE_PATH"` isn't `""`, but the value of that variable is (at least on our system).
In my investigation I got things to run by doing 

`include_list = [i for i in include_list if e.subst( i ) != ""]`

instead of what 1a6f3f4 proposes.

Btw., thanks for your efforts, Eric! :) It's awesome that you've taken this on. 

Just a quick note regarding review workflow: We at IE try to get into the habit of addressing John's feedback by responding with the commit hash that addresses it and then we let _him_ resolve the conversation once he's happy. I've unresolved the conversation for now so that we can get this sorted :)

Thanks!

johnhaddon replied:
> I think that fix doesn't work because we don't need to check the entry in include_list but the path after substitution instead. "$FREETYPE_INCLUDE_PATH" isn't "", but the value of that variable is (at least on our system).
In my investigation I got things to run by doing
>
>`include_list = [i for i in include_list if e.subst( i ) != ""]`

Good catch Matti - your suggested solution sounds good too. Could you add that please Eric?

ericmehl replied:
Thanks for taking a look at it Matti, that does make more sense. I've resolved that with 000ab6bfdcbcdb3f1180ea73b5daac5b6d54d363

And thanks for the note about review workflow!

johnhaddon replied:
Thanks Eric! Matti, could you give this one more test at your end, and merge if all seems well?

mattigruener replied:
I did, see comment below. Thanks for the fix, Eric!

ericmehl commented:
> Thanks for the update Eric, sorry it took me so long to get back to it. I really do want to see the Windows support merged just as soon as possible, but have been juggling other priorities for my Linux-based paymasters.

No problem at all, this is great timing actually - I have a solid amount of time to focus on Gaffer for a while. I'll get into these today.

> On a related topic, we've set up additional continuous integration for Gaffer using Azure Pipelines, and it seems to be working pretty well. One nice benefit is that it also has Windows support, so I'm intending to get the Cortex tests moved over to Azure as well. If we do that, do you see any reason why we couldn't use Azure instead of Appveyor for Windows testing?

I don't know of anything preventing that, it would be nice to have them on the same system. I'll get that set up too and we'll be cruising!

ericmehl commented:
Thanks a lot for the review John! I ticked off those remaining items, if there's anything else you spot or if these changes aren't quite right just let me know.

ericmehl commented:
Trying a close / reopen for single test failure

ericmehl commented:
Checks are finally passing! Apologies if I spammed subscribers with the opening/closing.

mattigruener commented:
Hey guys,

I've continued testing the build on my IE system today and I ran into a handful of other problems. I'm going to list the compiler errors and the diff that makes them go away below. Regarding the diff, I'm not sure these are the best approaches to fixing the issues and I'd like @ericmehl and @johnhaddon to take a look as I'm not too familiar with how things are generally structured in the SConstruct file. It looks to me like there have been some changes to how environments are inherited/cloned and that might have messed up some of the include paths?

I'm surprised that the VDB and the USD issues didn't come up in your tests, John? GafferHQ doesn't run the DCC specific tests, I guess, but the other ones should've cropped up?

Regarding the one I've listed as an RV issue: In our buildAll script for cortex we're running a last build for `APP=rv` and it looks like a previous commit has broken that. What seems to happen is that we're running that build with _Boost 1.55.0_ while most other things are compiled against _Boost 1.61.0_. [This is the commit ](https://github.com/ImageEngine/cortex/commit/d0731a3e9f0757760e618cb5a68d74e0651273b9) that breaks compatibility as far as I can see. By including `ExceptionBinding.inl` we pull in `boost/make_unique.hpp` that _Boost 1.55.0_ doesn't seem to ship with? Are we ok with breaking compatibility like that? I'm not entirely sure why we compile against the older boost, but it's potentially just because RV isn't good at adhering to VFX platform standards?

Ignoring the Boost problem, with my fixes in place, I can compile the code in this PR on our system.

Regarding commit history, are we going to remove the empty commits at the bottom of the commit stack? Shall we merge the two commits touching the previously discussed `include_list` filtering?

Thanks guys - looks like we're really close on this one which is exciting :)

Matti.

### Compiler Errors

USD Issue
```
/software/apps/gcc/4.8.3/cent7.x86_64/bin/g++ -o contrib/IECoreUSD/src/IECoreUSD/USDScene.os -c -pipe -Wall -pthread -isystem /software/tools/include/cent7.x86_64/boost/1.61.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/apps/tbb/4.4/cent7.x86_64/gcc/4.8.3/include -isystem /software/tools/include/cent7.x86_64/blosc/1.7.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -std=c++11 -fvisibility=hidden -Werror -DNDEBUG -DBOOST_DISABLE_ASSERTS -O3 -DBOOST_SIGNALS_NO_DEPRECATION_WARNING -Wno-deprecated -DBUILD_COMPONENT_SRC_PREFIX= -DBUILD_OPTLEVEL_DEV -isystem /software/apps/usd/18.09/cent7.x86_64/gcc/4.8.3/cortex/10/include -fPIC -DIE_CORE_MAJORVERSION=10 -DIE_CORE_MINORVERSION=0 -DIE_CORE_PATCHVERSION=0 -DBOOST_FILESYSTEM_VERSION=3 -Iinclude -Icontrib/IECoreUSD/src contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp
In file included from /software/tools/include/cent7.x86_64/boost/1.61.0/boost/python/detail/prefix.hpp:13:0,
                 from /software/tools/include/cent7.x86_64/boost/1.61.0/boost/python/object_fwd.hpp:8,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/base/tf/pyObjWrapper.h:32,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/base/vt/value.h:32,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/base/vt/keyValue.h:29,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/base/vt/dictionary.h:33,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/usd/sdf/types.h:64,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/usd/usd/primData.h:33,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/usd/usd/object.h:32,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/usd/usd/prim.h:32,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/usd/usd/schemaBase.h:29,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/usd/usd/apiSchemaBase.h:31,
                 from /software/apps/usd/18.09/cent7.x86_64/cortex/10/include/pxr/usd/usd/collectionAPI.h:31,
                 from contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp:51:
/software/tools/include/cent7.x86_64/boost/1.61.0/boost/python/detail/wrap_python.hpp:50:23: fatal error: pyconfig.h: No such file or directory
 # include <pyconfig.h>
                       ^
compilation terminated.
```
VDB Issue
```
/software/apps/gcc/4.8.3/cent7.x86_64/bin/g++ -o src/IECoreVDB/bindings/GafferVDBModule.os -c -pipe -Wall -pthread -isystem /software/tools/include/cent7.x86_64/boost/1.61.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/apps/tbb/4.4/cent7.x86_64/gcc/4.8.3/include -isystem /software/tools/include/cent7.x86_64/blosc/1.7.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -std=c++11 -fvisibility=hidden -Werror -DNDEBUG -DBOOST_DISABLE_ASSERTS -O3 -DBOOST_SIGNALS_NO_DEPRECATION_WARNING -isystem /software/apps/python/2.7/cent7.x86_64/gcc/4.8.3/include/python2.7 -Wno-strict-aliasing -fPIC -DIE_CORE_MAJORVERSION=10 -DIE_CORE_MINORVERSION=0 -DIE_CORE_PATCHVERSION=0 -DBOOST_FILESYSTEM_VERSION=3 -DBOOST_PYTHON_MAX_ARITY=20 -Iinclude src/IECoreVDB/bindings/GafferVDBModule.cpp
In file included from src/IECoreVDB/bindings/GafferVDBModule.cpp:42:0:
include/IECoreVDB/VDBObject.h:50:29: fatal error: openvdb/openvdb.h: No such file or directory
 #include "openvdb/openvdb.h"
                             ^
compilation terminated.
```
Nuke Issue
```
/software/apps/gcc/4.8.3/cent7.x86_64/bin/g++ -o src/IECoreNuke/bindings/FnOpHolderBinding.os -c -pipe -Wall -pthread -isystem /software/tools/include/cent7.x86_64/boost/1.61.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/apps/tbb/4.4/cent7.x86_64/gcc/4.8.3/include -isystem /software/tools/include/cent7.x86_64/blosc/1.7.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -std=c++11 -fvisibility=hidden -Werror -DNDEBUG -DBOOST_DISABLE_ASSERTS -O3 -DBOOST_SIGNALS_NO_DEPRECATION_WARNING -isystem /software/apps/python/2.7/cent7.x86_64/gcc/4.8.3/include/python2.7 -Wno-strict-aliasing -fPIC -DIE_CORE_MAJORVERSION=10 -DIE_CORE_MINORVERSION=0 -DIE_CORE_PATCHVERSION=0 -DBOOST_FILESYSTEM_VERSION=3 -DBOOST_PYTHON_MAX_ARITY=20 -isystem /software/apps/python/2.7/cent7.x86_64/gcc/4.8.3/include/python2.7 -Iinclude src/IECoreNuke/bindings/FnOpHolderBinding.cpp
In file included from include/IECoreNuke/ParameterisedHolder.h:38:0,
                 from include/IECoreNuke/OpHolder.h:38,
                 from src/IECoreNuke/bindings/FnOpHolderBinding.cpp:37:
include/IECoreNuke/ObjectKnob.h:43:27: fatal error: DDImage/Knobs.h: No such file or directory
 #include "DDImage/Knobs.h"
                           ^
compilation terminated.
```
Maya Issue
```
/software/apps/gcc/4.8.3/cent7.x86_64/bin/g++ -o src/IECoreMaya/bindings/FromMayaInstancerConverterBinding.os -c -pipe -Wall -pthread -isystem /software/tools/include/cent7.x86_64/boost/1.61.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/apps/tbb/4.4/cent7.x86_64/gcc/4.8.3/include -isystem /software/tools/include/cent7.x86_64/blosc/1.7.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -std=c++11 -fvisibility=hidden -Werror -DNDEBUG -DBOOST_DISABLE_ASSERTS -O3 -DBOOST_SIGNALS_NO_DEPRECATION_WARNING -isystem /software/apps/python/2.7/cent7.x86_64/gcc/4.8.3/include/python2.7 -Wno-strict-aliasing -DIECoreMaya_EXPORTS -fPIC -DIE_CORE_MAJORVERSION=10 -DIE_CORE_MINORVERSION=0 -DIE_CORE_PATCHVERSION=0 -DBOOST_FILESYSTEM_VERSION=3 -DBOOST_PYTHON_MAX_ARITY=20 -D_BOOL -DREQUIRE_IOSTREAM -isystem /software/apps/python/2.7/cent7.x86_64/gcc/4.8.3/include/python2.7 -DLINUX -Iinclude src/IECoreMaya/bindings/FromMayaInstancerConverterBinding.cpp
In file included from include/IECoreMaya/FromMayaDagNodeConverter.h:38:0,
                 from include/IECoreMaya/FromMayaInstancerConverter.h:38,
                 from src/IECoreMaya/bindings/FromMayaInstancerConverterBinding.cpp:37:
include/IECoreMaya/FromMayaObjectConverter.h:43:32: fatal error: maya/MObjectHandle.h: No such file or directory
 #include "maya/MObjectHandle.h"
                                ^
compilation terminated.
```
RV Issues
```
/software/apps/gcc/4.8.3/cent7.x86_64/bin/g++ -o src/IECorePython/HeaderGeneratorBinding.os -c -pipe -Wall -pthread -isystem /software/tools/include/cent7.x86_64/boost/1.55.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0 -isystem /software/apps/tbb/4.4/cent7.x86_64/gcc/4.8.3/include -isystem /software/tools/include/cent7.x86_64/blosc/1.7.0 -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -isystem /software/tools/include/cent7.x86_64/OpenEXR/2.2.0/OpenEXR -std=c++11 -fvisibility=hidden -Werror -DNDEBUG -DBOOST_DISABLE_ASSERTS -O3 -DBOOST_SIGNALS_NO_DEPRECATION_WARNING -isystem /software/apps/python/2.7/cent7.x86_64/gcc/4.8.3/include/python2.7 -Wno-strict-aliasing -DIECorePython_EXPORTS -fPIC -DIE_CORE_MAJORVERSION=10 -DIE_CORE_MINORVERSION=0 -DIE_CORE_PATCHVERSION=0 -DBOOST_FILESYSTEM_VERSION=3 -DBOOST_PYTHON_MAX_ARITY=20 -Iinclude src/IECorePython/HeaderGeneratorBinding.cpp
In file included from include/IECorePython/ExceptionBinding.h:71:0,
                 from src/IECorePython/CancellerBinding.cpp:40:
include/IECorePython/ExceptionBinding.inl:38:33: fatal error: boost/make_unique.hpp: No such file or directory
 #include "boost/make_unique.hpp"
                                 ^
compilation terminated.
```
### Diff
```
@@ -1924,6 +1924,8 @@ vdbEnv.Append( **vdbEnvAppends )
 
 vdbPythonModuleEnv = corePythonModuleEnv.Clone( IECORE_NAME="IECoreVDB" )
 
+appendSystemDirectory( vdbPythonModuleEnv, "$VDB_INCLUDE_PATH" )
+
 vdbPythonModuleEnv.Append( **vdbEnvAppends )
 
 vdbSources = sorted( glob.glob( "src/IECoreVDB/*.cpp" ) )
@@ -2236,6 +2238,8 @@ mayaEnv.Append( SHLINKFLAGS = pythonEnv["PYTHON_LINK_FLAGS"].split() )
 mayaPythonModuleEnv = pythonModuleEnv.Clone( **mayaEnvSets )
 mayaPythonModuleEnv.Append( **mayaEnvAppends )
 
+appendSystemDirectory( mayaPythonModuleEnv, "$MAYA_ROOT/include" )
+
 mayaPluginEnv = mayaEnv.Clone( IECORE_NAME="ieCore" )
 
 haveMaya = False
@@ -2439,6 +2443,7 @@ if env["PLATFORM"] == "darwin" :
 
 nukePythonModuleEnv = pythonModuleEnv.Clone( IECORE_NAME = "IECoreNuke" )
 nukePythonModuleEnv.Append( **nukeEnvAppends )
+appendSystemDirectory( nukePythonModuleEnv, ["$NUKE_ROOT/include"] )
 
 nukePluginEnv = nukeEnv.Clone( IECORE_NAME="ieCore" )
 
@@ -3049,6 +3054,7 @@ usdEnvAppends = {
 
 usdEnv.Append( **usdEnvAppends )
 appendSystemDirectory( usdEnv, "$USD_INCLUDE_PATH")
+appendSystemDirectory( usdEnv, "$PYTHON_INCLUDE_PATH")
 
 usdPythonModuleEnv = pythonModuleEnv.Clone( IECORE_NAME = "IECoreUSD" )
 usdPythonModuleEnv.Append( **usdEnvAppends )
```

johnhaddon commented:
Thanks for all the testing Matti!

> Regarding the diff, I'm not sure these are the best approaches to fixing the issues and I'd like @ericmehl and @johnhaddon to take a look as I'm not too familiar with how things are generally structured in the SConstruct file. It looks to me like there have been some changes to how environments are inherited/cloned and that might have messed up some of the include paths?

Ack, sorry, I should have caught this. The general pattern we had before was this :

```
mayaEnvAppends = {
    collectionOfImportantSettingsForCompilingMayaStuff
}
...
mayaEnv = env.Clone( **mayaEnvSets )
mayaEnv.Append( **mayaEnvAppends )
...
mayaPythonModuleEnv = pythonModuleEnv.Clone( **mayaEnvSets )
mayaPythonModuleEnv.Append( **mayaEnvAppends )
```

So we were sort of "mixing in" the maya-specific settings into clones of two core environments, one for C++, and one for Python. Because this PR lifts the include paths out of `mayaEnvAppends`, the python environment no longer gets them. Your fix works, but I think it would be preferable to not repeat ourselves if possible, and keep everything in `mayaEnvAppends()`. Eric, could we refactor `appendSystemDirectory()` so it doesn't modify an environment but _returns_ a list of `CXXFLAGS` so that we can do this instead?

```
mayaEnvAppends = {
    CXXFLAGS = [
	"-DIECoreMaya_EXPORTS",
     ] + formatSystemIncludes( 
           "$MAYA_ROOT/include"
     )
}
```

Will that work? If it does, it has the nice side effect of not needing to maintain near-identical `append()` and `prepend()` variants of the function...

> I'm surprised that the VDB and the USD issues didn't come up in your tests, John? GafferHQ doesn't run the DCC specific tests, I guess, but the other ones should've cropped up?

I think the difference is that GafferHQ/dependencies builds everything into a single root. So we have a lot more leeway when it comes to getting the include paths right for Cortex - we don't need to specify all the individual VDB/USD paths etc, because `$GAFFER_ROOT/include` catches everything anyway.

> Regarding the one I've listed as an RV issue...what seems to happen is that we're running that build with Boost 1.55.0 while most other things are compiled against Boost 1.61.0.

I've worked around that with this PR : https://github.com/ImageEngine/cortex/pull/965. I've also gone ahead and made a Cortex release with it in, so I can get GafferHQ/dependencies updated in preparation for Gaffer 0.54.

> Regarding commit history, are we going to remove the empty commits at the bottom of the commit stack? Shall we merge the two commits touching the previously discussed include_list filtering?

Sounds good to me. While we're nitpicking, perhaps we could also lose the `snake_case` and have a nice bit of `camelCase`?

Thanks for sticking with us on this Eric! I think we're nearly there! On a related note, this week I finally purged GafferHQ/dependencies off its remaining shell scripts, so everything now builds via the config-based system. Hopefully that gets us a little closer to using GafferHQ/dependencies to manage the Windows builds too...

Cheers...
John

ericmehl commented:
> Will that work? If it does, it has the nice side effect of not needing to maintain near-identical `append()` and `prepend()` variants of the function...

The new prepend / append functions operate on different variables (CPPPATH on Windows, CXXFLAGS otherwise) but I can compress them into one formatSystemIncludes and have it return a dictionary to join with mayaEnvAppends and similar.

> Sounds good to me. While we're nitpicking, perhaps we could also lose the `snake_case` and have a nice bit of `camelCase`?

Got it!

> Thanks for sticking with us on this Eric! I think we're nearly there! On a related note, this week I finally purged GafferHQ/dependencies off its remaining shell scripts, so everything now builds via the config-based system. Hopefully that gets us a little closer to using GafferHQ/dependencies to manage the Windows builds too...

My pleasure, I know this is a pretty big change and thanks to you and Matti and others for working with me on it. I've been tinkering with the GafferDependencies this week as well trying to fix an Appleseed crash so I'll get your latest pulled in and sync up the Windows build while I'm in that mindset.

Updates on this PR coming shortly!

johnhaddon commented:
> The new prepend / append functions operate on different variables (CPPPATH on Windows, CXXFLAGS otherwise) but I can compress them into one formatSystemIncludes and have it return a dictionary to join with mayaEnvAppends and similar.

I was assuming that you could use CXXFLAGS for Windows too, on the presumption that CPPPATH is just a convenience that then formats something like "-I<foo>" into the compiler flags. Is that not the case?

ericmehl commented:
> I was assuming that you could use CXXFLAGS for Windows too, on the presumption that CPPPATH is just a convenience that then formats something like "-I" into the compiler flags. Is that not the case?

I had been hoping that too but Scons does some more logic for cross-platform compatibility that makes this not work on Windows. Their documentation says:
> Don't explicitly put include directory arguments in CCFLAGS or CXXFLAGS because the result will be non-portable and the directories will not be searched by the dependency scanner.

and when I tried a while back to see how true that is it turns out they mean it - Windows spazzes out if includes are put into the CXXFLAGS. I think Scons puts CXXFLAGS into the command line as-is whereas it formats CPPPATH according to the platform. If we try and skirt around that we get into all kinds of messes with handling Windows vs posix throughout the script that 

> I was assuming that you could use CXXFLAGS for Windows too, on the presumption that CPPPATH is just a convenience that then formats something like "-I" into the compiler flags. Is that not the case?

Ah yes, we can use CXXFLAGS on Windows. Scons does a little more with CPPPATH than just plugging it in, but since we are doing the cross-platform logic ourselves it's safe to go directly to CXXFLAGS.

ericmehl commented:
The latest comments should be addressed by 77c3899. @johnhaddon and @mattigruener, can you please give those a try to see if it fixes your IE builds?

I also reformatted a couple of the modules that were not using the `moduleEnvAppends` dictionary format, let me know if that's undesirable and I can change them back.

johnhaddon commented on SConstruct:
```diff
@@ -1041,15 +1041,18 @@ env.Append(
-def includeSystemDirectory( e, include_list ):
-	if type(include_list) != list:
-		include_list = [include_list]
+def formatSystemIncludes( e, includeList ) :
```
Since this function isn't intended to modify an environment any more, we should remove the `e` argument. I see you're using it internally to get "PLATFORM", but we could do that with `sys.platform` or `Environment()["PLATFORM"]`.

ericmehl replied:
Yeah, I was waffling on that one. I'm also using it to check against empty include directories using `e.subst( i )`. I pass the specific module's environment such as `formatSystemIncludes( mayaEnv, includes)` in case in the future there are module-specific substitutions.

Do you think I should just use `env.subst( i )` instead so we can eliminate that parameter?

johnhaddon replied:
> I'm also using it to check against empty include directories using e.subst( i ).

Ah, yes, I missed that. Let's leave it as you have it - thanks for the explanation.

mattigruener commented on SConstruct:
```diff
@@ -3025,7 +3041,7 @@ usdEnvAppends = {
 		"-DBUILD_COMPONENT_SRC_PREFIX=",
 		"-DBUILD_OPTLEVEL_DEV",
-	],
+	] + formatSystemIncludes( usdEnv, "$USD_INCLUDE_PATH" ),
```
We'll also need `$PYTHON_INCLUDE_PATH` included here.

ericmehl replied:
Great, included in d5287b7

mattigruener replied:
LGTM, thanks :)

ericmehl commented:
I'll close and reopen to try and get that last test passing that fails intermittently.

johnhaddon commented:
Where are we at with this @mattigruener? Good to merge now? @andrewkaufman, and final comments?

andrewkaufman commented:
I've not been following this at all unfortunately, and I doubt I'd have time to review before the Gaffer 0.54 release... if its working for GafferHQ and working for a full IE release (@mattigruener have you tested that with all the dccs?) then I say merge (though it'd be nice to get rid of those early branch merge commits)

mattigruener commented:
Yes, good to merge as far as I'm concerned. Thanks for your work on this, Eric! @andrewkaufman?

ericmehl commented:
My pleasure!

> though it'd be nice to get rid of those early branch merge commits

I think I sorted out how to drop those commits, I'm happy to force-push them here or leave as-is, let me know what you like best.

boberfly commented:
Awesome

andrewkaufman commented:
Just a heads up, in addition to #972, we have a crash when building Cortex for RV (ie. with an ancient set of dependencies). We'll have to sort that out before we merge.

andrewkaufman commented:
It looks to be boost filesystem related fwiw

andrewkaufman commented:
Seems to have been an anomaly in Matti's build. RV is working fine for me with this PR rebased onto master. I verified our other build iterations as well.

By the way, doing that rebase removed the merge commits Eric, so I think all you'd need todo before we merge is:

```
git checkout master
git pull <ImageEngine remote> master
git checkout scons_windows
git rebase master
```

verify the commit history, and then

```
git push <Hypothetical remote> scons_windows --force
```

ericmehl commented:
Strange... it's failing because it can't find scons. I'll try a close and reopen to trigger a new build.

ericmehl commented:
I rebased as Andrew described and it's passing fine on my non-PR Travis builds and Windows Appveyor builds, but it's failing here because it thinks scons is not installed. Is that happening on any other builds you've been running? The Travis documentation still lists scons as being part of Linux images.

johnhaddon commented:
This is happening on other PRs too - seems to be a Travis issue. It's not affecting the Gaffer PRs though, and the main difference between setups seems to be that Cortex has `sudo:false` whereas Gaffer doesn't specify it at all, and Gaffer has `dist: trusty`. Could you try editing `.travis.yml` to match Gaffer in those respects and repush?

ericmehl commented:
Looks like your fix worked John, thanks for the help. Getting a single failure like I've seen before, so giving it a close and reopen to try and fix.

Just curious, are these seemingly random test failures something you all see a lot too? Or is it something I can remedy in my builds?

andrewkaufman commented:
> Just curious, are these seemingly random test failures something you all see a lot too? Or is it something I can remedy in my builds?

Yes we get them too... sometimes it a problem with a particular test (or a sign of an intermittent bug). But more often than not, its Travis being Travis. That's part of the motivation for migrating to Azure to be honest.

Travis seems happy now, but @johnhaddon still has a change request showing, though I'm not sure if that's just stale or what. I'll wait for him to dismiss it before merging.

ericmehl commented:
I realized I had a misleading comment about the -isystem include change when getting into a similar conversion of Gaffer to Scons so I figured I'd update this since it hasn't been merged yet.

Hopefully it doesn't make more work for your approval workflow, apologies if it does.

ericmehl commented:
Restarting for an errant test failure

andrewkaufman commented:
I'll merge this now and it should make it into the next tag (10.0.0-a61). Thanks for the hard work @ericmehl !

johnhaddon commented:
> I'll merge this now and it should make it into the next tag (10.0.0-a61). Thanks for the hard work @ericmehl !

🎉 Thanks Eric!

ericmehl commented:
Thanks for all your help guys! Great to see it merged in :)

andrewkaufman commented:
@ericmehl where/how can we see the appveyor results for each PR? I was expecting them to show up in the Check section alongside Travis, but I guess that's not the case?

ericmehl commented:
It should show up in the Check section - maybe the organization settings are not set up? Try this Appveyor url: [https://ci.appveyor.com/authorizations](https://ci.appveyor.com/authorizations) and under Github follow the "organization access" link to the Image Engine Github organization settings.

Does Appveyor show up as approved in that section?

andrewkaufman commented:
It wasn't installed for the Org yet, but now is with access granted for Cortex. I force pushed to #937, but that doesn't seem to have kicked off any new checks...

andrewkaufman commented:
Ok, it seems to be queued now, after closing and re-opening the PR. So are force-pushes not enough to trigger Appveyor I take it?

ericmehl commented:
I haven't used it much with pull requests so I'm not 100% sure how it handles the close and re-open. It's been building my force-pushes to a regular branch though. Maybe it handles PR pushes differently?

I do have a working Azure Pipelines configuration I just made a PR for that I made in anticipation of the Linux CI builds going to that platform. Maybe we drop Appveyor in favor of that?

andrewkaufman commented:
Yep, lets migrate to Azure soon. Or maybe GitHub Actions.

andrewkaufman commented:
@ericmehl do you have a recommended version of visual studio that works best with Cortex?

ericmehl commented:
Yes, I've been building with Visual Studio 2017 (Community a.k.a. free edition but it should be the same as the Enterprise version). 

I'm pretty sure Visual Studio 2015 won't work without some possibly significant work, but it's been a long time since I tried it. I haven't tested on the new 2019 edition yet.

boberfly commented:
Yep start with msvc2017, there is a critical template bug with msvc2015 which causes issues when trying to compile the boost python bindings, and the workaround that USD uses doesn't work on base classes which Gaffer/Cortex has.
